python/lsst/ip/isr/binExposureTask.py

```diff
@@ -0,0 +1,202 @@
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = ["BinExposureTask",
+           "BinExposureConfig",
+           "binExposure"]
+
+import lsst.pipe.base as pipeBase
+import lsst.pipe.base.connectionTypes as cT
+import lsst.pex.config as pexConfig
+import lsst.afw.math as afwMath
+from lsst.utils.timer import timeMethod
+import lsst.afw.image as afwImage
+
+
+class BinExposureConnections(
+    pipeBase.PipelineTaskConnections,
+    dimensions=("instrument", "exposure", "detector"),
+    defaultTemplates={"inputName": "postISRCCD", "outputName": "postISRCCDBin"}
+):
+
+    inputExposure = cT.Input(
+        name="{inputName}",
+        doc="Input exposure to bin.",
+        storageClass="ExposureF",
+        dimensions=["instrument", "exposure", "detector"],
+    )
+    binnedExposure = cT.Output(
+        name="{outputName}",
+        doc="Binned exsposure.",
+        storageClass="ExposureF",
+        dimensions=["instrument", "exposure", "detector"],
+    )
+
+    def __init__(self, *, config=None):
+        """Customize the connections and storageClass for a specific
+        instance. This enables both to be dynamically set at runtime,
+        allowing BinExposureTask to work with different types of
+        Exposures such as postISRCCD, calexp, deepCoadd_calexp, etc.
+
+        Parameters
+        ----------
+        config : `BinExposureConfig`
+            A config for `BinExposureTask`.
+        """
+        super().__init__(config=config)
+        if config and config.exposureDimensions != self.inputExposure.dimensions:
+            self.dimensions.clear()
+            self.dimensions.update(config.exposureDimensions)
+            self.inputExposure = cT.Input(
+                name=self.inputExposure.name,
+                doc=self.inputExposure.doc,
+                storageClass=self.inputExposure.storageClass,
+                dimensions=frozenset(config.exposureDimensions),
+            )
+            self.binnedExposure = cT.Output(
+                name=self.binnedExposure.name,
+                doc=self.binnedExposure.doc,
+                storageClass=self.binnedExposure.storageClass,
+                dimensions=frozenset(config.exposureDimensions),
+            )
+        if config and config.exposureStorageClass != self.inputExposure.storageClass:
```

Comment:
Do you have any expected cases where the `storageClass` would need to change?

Reply:
I think this also would cause problems with the check on L193, as that will raise if the `storageClass` isn't an `Exposure`.  This does allow `ExposureD`, so maybe a comment here mentioning that although the `storageClass` can be changed, if it isn't some kind of `Exposure`, it's going to cause problems.

Reply:
I added this to ensure that the output storage class would always match the input storage class. While I could have the input and output storage classes as the generic `Exposure`, and it would be happy with any of the sub-types as input, any subtype information (I,F,D etc) would be lost on the output (at least that's my understanding).

I can certainly add the comment.